relay-server/src/actors/project_buffer.rs

```diff
@@ -189,11 +213,130 @@ impl Service for BufferService {
     }
 }
 
-impl Drop for BufferService {
+impl Drop for MemoryBufferService {
     fn drop(&mut self) {
         let count: usize = self.buffer.values().map(|v| v.len()).sum();
         if count > 0 {
             relay_log::error!("dropped queue with {} envelopes", count);
         }
     }
 }
+
+/// [`Buffer`] interface implementation backed by SQLite.
+#[derive(Debug)]
+pub struct SqliteBufferService {
+    db: Pool<Sqlite>,
+}
+
+impl SqliteBufferService {
+    /// Creates a new [`SqliteBufferService`] from the provide path to the sqlite database file.
+    pub async fn from_path(path: PathBuf) -> Result<Self, BufferError> {
+        let options = SqliteConnectOptions::new()
+            .filename(PathBuf::from("sqlite://").join(path))
+            .journal_mode(SqliteJournalMode::Wal)
+            .create_if_missing(true);
+
+        let db = SqlitePoolOptions::new().connect_with(options).await?;
+        Ok(Self { db })
+    }
+
+    /// Handles the enqueueing messages into the internal buffer.
+    async fn handle_enqueue(&self, message: Enqueue) -> Result<(), BufferError> {
+        let Enqueue {
+            key,
+            value: mut managed_envelope,
+        } = message;
+
+        managed_envelope.spool();
+        let envelope_bytes = managed_envelope.take_envelope().to_vec()?;
```

Comment:
`take_envelope()` is an ugly hack (that's on me). How about we introduce a `ManagedEnvelope::into_parts()` method for this use case? That could internally give back the permit.

Reply:
yeah, that would definitely make my life easier 😄